### ISSUE TITLE
C string expression not consistent across machines

### ISSUE THREAD

**Issue by tkarna** [firedrakeproject/firedrake#1024]
One Thetis convergence test (`test/momentumEq/test_v-viscosity_mes.py`) passes on my laptop but not on my desktop. Firedrake install is identical in both machines (as reported by `firedrake-status`). Laptop runs Ubuntu 16.04, desktop Ubuntu 16.10.

It turns out that the initial condition projected from a C string expression is different!
See the minimal example:

```python
from firedrake import *

lx = 7.0e3
ly = 5.0e3
area = lx*ly
depth = 40.0
v_viscosity = 5e-3

n_layers = 6
nx = 3
ny = 2
mesh2d = RectangleMesh(nx, ny, lx, ly)
mesh = ExtrudedMesh(mesh2d, n_layers, depth/n_layers)

p1dg_v = VectorFunctionSpace(mesh, 'DG', 1)
uv_3d = Function(p1dg_v)

t = 100.0  # initial simulation time

# c-string expression
# NOTE L2 norm on my laptop is  35997.937407009391
# NOTE L2 norm on my desktop is 35697.265219521272 (bad enough to affect convergence)
ana_sol_expr = '0.5*(u_max + u_min) - 0.5*(u_max - u_min)*erf((x[2] - z0)/sqrt(4*D*t))'
t_const = Constant(t)
ana_uv_expr = Expression((ana_sol_expr, 0.0, 0.0), u_max=1.0, u_min=-1.0, z0=depth/2.0, D=v_viscosity, t=t_const)

uv_3d.project(ana_uv_expr)
print('Initial norm uv     {:.12f}'.format(norm(uv_3d)))

# ufl expression
# NOTE L2 norm on my laptop is  36097.470155039759
# NOTE L2 norm on my desktop is 36097.470155039759
u_max=1.0
u_min=-1.0
xyz = SpatialCoordinate(mesh)
z0=depth/2.0
D=v_viscosity
t_const = Constant(t)
ana_sol_expr = 0.5*(u_max + u_min) - 0.5*(u_max - u_min)*erf((xyz[2] - z0)/sqrt(4*D*t_const))
ana_uv_expr = as_vector((ana_sol_expr, 0.0, 0.0))

uv_3d.project(ana_uv_expr)
print('Initial norm uv     {:.12f}'.format(norm(uv_3d)))
```

If I replace the C string expression with a UFL expression I get the same solution on both machines. Obviously the solution is to remove all C string expressions. But if someone still uses C string expressions, you might encounter unexpected behaviour.

Could someone running Ubuntu 16.04 or 16.10 please try to reproduce? I think the only difference is the OS but I'm not entirely sure...

**Comment by tkarna** [comment]
`firedrake-status` on both machines is
```
Firedrake Configuration:
    slope: False
    package_manager: True
    slepc: False
    adjoint: True
    minimal_petsc: False
    petsc_int_type: int32
    mpicc: mpicc
    disable_ssh: False
    show_petsc_configure_options: False
    packages: []
    honour_pythonpath: False
    developer: True
Additions:
    None
Environment:
    PYTHONPATH: None
    PETSC_ARCH: None
    PETSC_DIR: None
Status of components:
---------------------------------------------------------------------------
|Package             |Branch                        |Revision  |Modified  |
---------------------------------------------------------------------------
|COFFEE              |master                        |87e5078   |False     |
|FInAT               |master                        |2bd9d80   |False     |
|PyOP2               |master                        |afff500   |False     |
|dolfin-adjoint      |master                        |fa4ee04   |False     |
|fiat                |master                        |184601a   |False     |
|firedrake           |master                        |33995cc   |False     |
|glpk                |master                        |45156a5   |True      |
|h5py                |firedrake                     |c1e4a81   |False     |
|libadjoint          |master                        |aeabc2e   |False     |
|libspatialindex     |master                        |34fd22b   |True      |
|petsc               |firedrake                     |e86e009   |False     |
|petsc4py            |firedrake                     |325db39   |False     |
|tsfc                |master                        |15160e6   |False     |
|ufl                 |master                        |c5eb7fb   |False     |
---------------------------------------------------------------------------
```

**Comment by wence-** [comment]
Can you inspect the code that is generated on the two different laptops to project the expression?  What happens is that the Expression is interpolated into some space, and then projected into the target space.

In the UFL expression case, that is projected directly.

**Comment by tkarna** [comment]
OK, so if I only project the C string expression, I get 5 pyop2 kernels. Between the two machines the file names differ but the generated C code is identical. The compiled *.so files differ.

Here's an example of a kernel
- Ubuntu 16.04 https://gist.github.com/tkarna/3f4d1f83b61378ff97bf0d07e0d5c44d
- Ubuntu 16.10 https://gist.github.com/tkarna/e6d3e67b8cdafdbc6471ad25cd5e79c8

FWIW, Ubuntu 16.04 has Python 2.7.12 and gcc 5.4.0., while Ubuntu 16.10 comes with Python 2.7.12+ and gcc 6.2.0. Could this be a compiler issue?

**Comment by tkarna** [comment]
Hmm if I replace the *.so binaries on my desktop (Ubuntu 16.10) by those generated on my laptop (Ubuntu 16.04) (renaming them appropriately), I get the same result as on my laptop. So it indeed seems that the difference is in the *.so files...

**Comment by wence-** [comment]
OK, thanks, I'll try and boot 16.10 and see if I can figure out what's going on.

**Comment by wence-** [comment]
FWIW, what do you get if you `interpolate` the expressions rather than projecting.  Does stuff come out looking the same?

**Comment by tkarna** [comment]
With `interpolate` I get the same result on both machines.

I guess I could try using `gcc-5` Ubuntu 16.10. Just setting `CC=gcc-5` didn't work because the JIT compilation fails (`mpi.h: No such file or directory`)...

**Comment by miklos1** [comment]
Try `MPICH_CC=gcc-5` or `OMPI_CC=gcc-5`.

**Comment by wence-** [comment]
Reduced testcase:
```python
from firedrake import *

mesh2d = UnitSquareMesh(1, 1, quadrilateral=False)
mesh = ExtrudedMesh(mesh2d, 1)

d = 1
V0 = FunctionSpace(mesh, 'DG', d)
uv_3d = Function(V0)

ana_sol_expr = 'x[2]'
ana_uv_expr = Expression(ana_sol_expr)
V = FunctionSpace(mesh, "DG", d+1)
f = interpolate(ana_uv_expr, V)

u = TrialFunction(V0)
v = TestFunction(V0)

solve(u*v*dx == f*v*dx, uv_3d, options_prefix="")

print norm(f), norm(uv_3d), norm(assemble(f*v*dx))
```

Produces:

`0.57735026919 0.57735026919 0.0424312564348` on 16.04

`0.57735026919 0.167066975227 0.0424312564348` on 16.10

Hmmm.

**Comment by dorugeber** [comment]
Analytically, we expect 0.57735026919, right?

We are taking the norm of 'z' in a unit cube, and sqrt(int_0^1 z^2 dz) = sqrt(1/3) = 1/sqrt(3) = 0.5773...

**Comment by tkarna** [comment]
@wence- I get the same results. Interesting. So the bug is in the projection, not in the c string expression itself; ufl expression gives the same error.
@dorugeber Yes, I think that's correct. Projecting the ufl expression itself gives the right answer.

**Comment by wence-** [comment]
Plausibly a C compiler bug?

This snippet (which should produce `24`) produces `12` on gcc-6 with `-O3` turned on.

I don't think it contains any undefined behaviour.

```c
#include <stdio.h>

void compute_integral (double  *A,
                       double  w_1[18] )
{
    double  t33[2][6] = {0};
    double  t43[2] = {0};
    const double  t31[2][2]  = {{1, 1},
                                {1, 1}};
    const double  t32[2][3]  = {{0, 0, 1},
                                {0, 0, 1}};

    for (int ip_1 = 0; ip_1 < 2; ++ip_1) {

        for (int i_0 = 0; i_0 < 6; ++i_0) {
            t33[ip_1][i_0] = (w_1[i_0*3] * t32[ip_1][0]) + (w_1[i_0*3+2] * t32[ip_1][2]);
        }
        t43[ip_1] = (t31[ip_1][0]) + (t31[ip_1][1]);
    }

    for (int ip_1 = 0; ip_1 < 2; ++ip_1) {
        double  t47 = 0.0;

        for (int i_0 = 0; i_0 < 6; ++i_0) {
            t47 += t33[ip_1][i_0];
        }
        A[0] += t47*t43[ip_1];
    }
}

int main()
{
    double A = 0;
    double w_1[18] = {0., 1.0, 1.0,
                      0., 1.0, 1.0,
                      0., 1.0, 1.0,
                      0., 1.0, 1.0,
                      0., 1.0, 1.0,
                      0., 1.0, 1.0};
    compute_integral(&A,
                     w_1);
    printf("%g\n", A);
    return 0;
}
```

**Comment by wence-** [comment]
Even smaller:

```c
int printf(const char * restrict, ...) __attribute__((__format__ (__printf__, 1, 2)));

void compute_integral (double  *A,
                       double  w_1[18] )
{
    double  t33[2][6] = {{0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
                         {0.0, 0.0, 0.0, 0.0, 0.0, 0.0}};
    double  t43[2] = {0.0, 0.0};
    double  t31[2][2]  = {{1.0, 1.0},
                          {1.0, 1.0}};
    double  t32[2][3]  = {{0.0, 0.0, 1.0},
                          {0.0, 0.0, 1.0}};

    for (int ip_1 = 0; ip_1 < 2; ++ip_1) {

        for (int i_0 = 0; i_0 < 6; ++i_0) {
            t33[ip_1][i_0] = (w_1[i_0*3] * t32[ip_1][0]) + (w_1[i_0*3+2] * t32[ip_1][2]);
        }
        t43[ip_1] = 2.0;
    }
    for (int i_0 = 0; i_0 < 6; ++i_0) {
        A[0] += t43[1]*t33[1][i_0];
    }
}

int main()
{
    double A = 0;
    double w_1[18] = {0., 1.0, 1.0,
                      0., 1.0, 1.0,
                      0., 1.0, 1.0,
                      0., 1.0, 1.0,
                      0., 1.0, 1.0,
                      0., 1.0, 1.0};
    compute_integral(&A, w_1);
    printf("%g\n", A);
    return 0;
}
```

**Comment by wence-** [comment]
Produces 0, not 12.

**Comment by wence-** [comment]
Let's see what happens here https://gcc.gnu.org/bugzilla/show_bug.cgi?id=79920

**Comment by wence-** [comment]
PyOP2 workaround fixes in OP2/PyOP2#523

**Comment by wence-** [comment]
Fix is in, WFM in 17.04.
